nc_compare/nc_compare.1.txt
nc_compare(1)
===========
:doctype: manpage


NAME
----
nc_compare - Compare netCDF files and report differences.


SYNOPSIS
--------
*nc_compare* ['OPTIONS'] 'LEFTFILE' 'RIGHTFILE'


DESCRIPTION
-----------

nc_compare(1) reads two netcdf files and compares their contents, printing
out differences in the dimensions, global attributes, variables, and
variable attributes.  The LEFTFILE and RIGHTFILE are taken as in diff(1),
and the output looks similar to a context diff of ncdump(1) output.  Unlike
diff(1), data values are also compared numerically, and some attributes are
ignored by default since they are expected to be different in different
files.  Also time coordinates are interpreted as times and reported as
different if they do not overlap exactly.

Numeric variables with identical names are compared statistically.  All
the values for each variable are read, missing values removed, and the mean
is computed.  The means of the two variables are used to compute an
absolute difference and a relative error between the variables.  Variables
are reported as different when their means differ according to the current
floating point comparison method.  Statistical differences are printed out
in a table following all the other output, with the variables ordered from
largest relative error to least.

Numeric attribute values are compared as numbers rather than as text.  So
floating point values do not have to be exactly equal to be considered not
different, they just have to compare as equal with the current floating
point comparison method.

By default, floating point values only have to be close to be considered
equal, even if they are not exactly equal.  For all other data types, the
values must be exactly equal.  There are two options for comparing floats, one
uses an error delta and the other compares Units in the Last Place.  See the
--delta and --ulps options for details.


OPTIONS
-------

*--showequal*::

    Show the text summary of every component in the netcdf files, even if
    they are equal between the two files.  Usually only the differences are
    reported.

*--ignore*='IGNORE'::

    Set a variable or attribute name to ignore when checking for
    differences.  By default nc_compare(1) ignores these names which
    typically appear as global attributes and vary for each generation of a
    data file:

    date_created, ProcessorURL, ProcessorRevision, NIDASrevision

    Passing an ignore option resets the list, so you can pass an empty or
    unused name to reset the ignore list and show everything, like so:

    `--ignore ''`

    Multiple ignore options can be used, each use adds a name to the ignore
    list.
    
*--delta*='DELTA'::

    Specify that floating point values must be within DELTA of each other
    to be considered equal.  The default error delta is 1e-6 for 32-bit
    float and 1e-10 for 64-bit float, but this option must be set to change
    the comparison method from the default ULPS method.

*--ulps*='ULPS'::

    Specify that floating point numbers must be within ULPS steps of each
    other to be considered equal.  One floating point number is advanced to
    the next representable floating point value towards the other number,
    and if the numbers are within ULPS steps of each other, they are close
    enough.  The default floating point comparison test uses 4 ULPS.

*--help*::

    Show usage information.

EXIT STATUS
-----------
*0*::
    Success

*N*::
    Failure.  The exit status is the number of differences found.


EXAMPLES
--------

.........................
> nc_compare HIPPO/BASE/HIPPO-5rf05.nc HIPPO/HIPPO-5rf05.nc
--- HIPPO/BASE/HIPPO-5rf05.nc (left)
+++ HIPPO/HIPPO-5rf05.nc (right)
variables:
!          float IRIG_Tdiff_LWO(Time) ;
 -                 SerialNumber = "F2DC001" ;
 +                 SerialNumber = "F2DC003" ;
!          float PS_VXL(Time) ;
 -                 actual_range = 183.317749, 1005.51453 ;
 +                 actual_range = 183.318161, 1005.51678 ;
 -                 CalibrationCoefficients = 0, 1.33322001 ;
 +                 CalibrationCoefficients = 0, 1.33322299 ;
   Variable              Left Mean     Right Mean  Abs Error  Rel Err (%)
   --------              ---------     ---------- ---------- ------------
   PS_VXL               629.370728     629.372131  -0.001404         0.00
19 differences.
.........................

The actual_range and CalibrationCoefficients attributes of the PS_VXL
variable differ by more than 4 ULPS, and so they are reported in the
output.  The '-' line shows the values from the LEFTFILE, while the '+'
line shows the RIGHTFILE.  The means for the variable also differ, and so
the statistics for that variable are reported in a table.  If the
`--showequal` option were enabled, then all the dimensions, variables,
attributes and statistics would be shown.

AUTHOR
------

nc_compare was originally written by Chris Webster for EOL aircraft data files,
then extended by Gary Granger to work with EOL ISS and ISFS data files.


nc_compare/nc_compare.cc
/*
-------------------------------------------------------------------------
OBJECT NAME:	nc_compare.cc

FULL NAME:	netCDF compare program

DESCRIPTION:	Compare two netCDF files and output the differnces.

COPYRIGHT:	University Corporation for Atmospheric Research, 2016
-------------------------------------------------------------------------
*/

#include "NcComparison.h"

#include <iostream>
#include <boost/program_options.hpp>

using std::cout;
using std::cerr;
namespace po = boost::program_options;

int
nc_compare(int argc, char *argv[])
{
  // Declare the supported options.
  po::options_description
    desc("nc_compare [options] primary_file secondary_file");
  desc.add_options()
    ("help", "Show this help message.")
    ("showequal", "Show equal objects as well as different.")
    ("showindex",
     "For vector values, report the indexes of the differences.")
    ("ignore", po::value<std::vector<std::string> >()->composing(),
     "Ignore attributes and variables with the given name.  "
     "Add * as suffix or prefix to match a substring.  "
     "Pass --ignore for each name to be ignored.")
    ("variable", po::value<std::vector<std::string> >()->composing(),
     "Select variables for comparison.  Add * as prefix or suffix "
     "or both match variables with the given substring.  "
     "All variables are selected by default.  "
     "If a variable matches one of the ignores, then it is ignored "
     "even if it matches one of the --variable strings.  "
     "Pass --variable for each substring to match.")
    ("override", po::value<std::vector<std::string> >()->composing(),
     "Override a global attribute in the left file specified in the form "
     "key=value.  "
     "Pass --override for each attribute to override.  Only "
     "global attribute names which exist in the left file will be "
     "overridden, all other overrides are ignored.")
    ("file", po::value<std::vector<std::string> >(), "Input files.")
    ("delta", po::value<double>(),
     "Error delta allowed for floating point values to be equal.  "
     "This is a floating point delta value, meaning "
     "the absolute difference between two numbers, no matter how large, must "
     "be less than the error delta.  To determine if floats of arbitrary "
     "range are close enough, use the ulps option.")
    ("ulps",
     po::value<int>()->default_value(compare_floating_point().getULPS()),
     "Number of Units in the Last Places (bits) "
     "in which floating point numbers can "
     "differ and still be considered equal.  This is the floating point "
     "comparison used if --delta or --ulps is not specified explicitly.  "
     "For example, ULPS=4 means two floats must be within 15 (2^4-1) steps "
     "of each other to compare as equal.  "
     "If ULPS is 0, then floats must be exactly equal.  "
     "The same ULPS setting is used for float and double types.")
    ("limit",
     po::value<int>()->default_value(ReportStyle::DEFAULT_REPORT_LIMIT),
     "Maximum number of differences to show in the report.  Once the limit "
     "is reached, no more differences are shown.")
    ("use-right-blanks",
     "Fill values in a variable on the right are used to filter "
     "out variable values on the left. In other words, values on the right "
     "which have been blanked out are also used to blank out the left, but "
     "only if the dimension sizes are the same.")
    ("minmax",
     "Report min and max in the statistics report.")
    ;

  po::positional_options_description p;
  p.add("file", 2);

  po::variables_map vm;
  po::store(po::command_line_parser(argc, argv).
	    options(desc).positional(p).run(), vm);
  po::notify(vm);

  if (vm.count("help")) {
    cout << desc << "\n";
    return 1;
  }

  std::vector<std::string> files;
  if (vm.count("file"))
    files = vm["file"].as<std::vector<std::string> >();
  if (files.size() != 2)
  {
    cerr << "Two input files must be specified.\n";
    exit(1);
  }

  ncopts = 0;
  putenv((char *)"TZ=UTC");	// All time calcs in UTC.

  NcCache left(files[0]);
  NcCache right(files[1]);

  std::vector<std::string> override;
  if (vm.count("override"))
  {
    override = vm["override"].as<std::vector<std::string> >();
    left.overrideGlobalAttributes(override);
  }

  CompareNetcdf ncdiff(&left, &right);
  ReportStyle& style = ncdiff.style();
  style.showEqual(vm.count("showequal") > 0);
  style.showIndex(vm.count("showindex") > 0);
  style.useRightBlanks(vm.count("use-right-blanks") > 0);
  style.showMinMax(vm.count("minmax") > 0);

  std::vector<std::string> ignores;
  if (vm.count("ignore"))
    ignores = vm["ignore"].as<std::vector<std::string> >();
  if (ignores.size())
  {
    ncdiff.ignore(ignores);
  }
  std::vector<std::string> selects;
  if (vm.count("variable"))
    selects = vm["variable"].as<std::vector<std::string> >();
  if (selects.size())
  {
    ncdiff.selectVariables(selects);
  }
  if (vm.count("limit"))
  {
    int limit = vm["limit"].as<int>();
    style.setReportLimit(limit);
  }
  compare_floating_point cfp;
  if (vm.count("ulps"))
  {
    int ulps = vm["ulps"].as<int>();
    cfp.setULPS(ulps);
    cfp.useULPS();
  }
  if (vm.count("delta"))
  {
    double delta = vm["delta"].as<double>();
    cfp.setDoubleDelta(delta);
    cfp.setFloatDelta(delta);
    cfp.useDelta();
  }
  ncdiff.setFloatComparator(cfp);
  ncdiff.compare();
  ncdiff.report(cout);
  int ndiffs = ncdiff.countDifferences();
  if (ndiffs)
  {
    std::cout << ndiffs << " differences.\n";
  }
  return ndiffs;
}


int
main(int argc, char *argv[])
{
  try {
    return nc_compare(argc, argv);
  }
  catch (const std::runtime_error& error)
  {
    // Typically IO errors.
    cerr << "Error: " << error.what() << std::endl;
    exit(1);
  }
  catch (const boost::program_options::error& error)
  {
    // Error parsing command-line options.
    cerr << error.what() << std::endl;
    cerr << "Use option --help to see usage info." << std::endl;
    exit(1);
  }
}
    
